### PR TITLE
[master] Finiteness checker

### PR DESCRIPTION
**Pull request by Alexsandruss**
# Description
NaN/Inf checker for Numeric Tables with base and AVX512 branches for float32/64 types.
Checker has two stages:

1. Compute sum of all values and check its finiteness
2. If sum isn't finite, check finiteness of all values.

Duplicate of pull request [373](https://github.com/intel/daal/pull/373) to 2020-mnt

### CODE REVIEW

**Comment by SmirnovEgorRu** [comment]
There are many things what can be improved, but let's move them in the next release.
For now it's approved, if it works on all compilers.

**Comment by SmirnovEgorRu** [comment]
The main concern what you can fix easily - fixing coding guidelines.

File: service/kernel/data_management/finiteness_checker.cpp [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+* limitations under the License.
+*******************************************************************************/
+
+#include "data_management/data/internal/finiteness_checker.h"
+#include "data_management/data/numeric_table.h"
+#include "services/env_detect.h"
+#include "externals/service_dispatch.h"
+#include "algorithms/threading/threading.h"
+#include "service_numeric_table.h"
+#include <chrono>
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: delete

Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Already in previous commit

**Comment by ShvetsKS** [comment]
make test

**Comment by Alexsandruss** [comment]
http://intel-ci.intel.com/ea4c0074-49dc-f18c-996a-8cdcd4b6d2c4

File: service/kernel/data_management/finiteness_checker.cpp [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+namespace daal
+{
+namespace data_management
+{
+namespace internal
+{
+typedef daal::data_management::NumericTable::StorageLayout NTLayout;
+
+void valuesAreFinite(float * dataPtr, bool * finitenessPrt, size_t n, bool allowNaN)
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: const float* dataPtr

File: service/kernel/data_management/finiteness_checker.cpp [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+            if (0x00000000u != (*uint32Ptr & 0x007fffffu) && allowNaN)
+                finitenessPrt[i] = true;
+            else
+                finitenessPrt[i] = false;
+        else
+            finitenessPrt[i] = true;
+    }
+}
+
+void valuesAreFinite(double * dataPtr, bool * finitenessPrt, size_t n, bool allowNaN)
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: const double* dataPtr

File: service/kernel/data_management/finiteness_checker.cpp [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+                finitenessPrt[i] = true;
+            else
+                finitenessPrt[i] = false;
+        else
+            finitenessPrt[i] = true;
+    }
+}
+
+template <typename DataType, daal::CpuType cpu>
+DataType computeSum(size_t nDataPtrs, size_t nElementsPerPtr, DataType ** dataPtrs)
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: const DataType **

File: service/kernel/data_management/finiteness_checker.cpp [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        if (layout == NTLayout::soa)
+            table.releaseBlockOfColumnValues(blockDescrPtr[i]);
+        else
+            table.releaseBlockOfRows(blockDescrPtr[i]);
+
+    *finiteness = valAreFinite;
+}
+
+template <typename DataType>
+DAAL_EXPORT bool allValuesAreFinite(NumericTable & table, bool allowNaN)
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: const NumericTable &

File: service/kernel/data_management/finiteness_checker.cpp [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    {
+        // SOA layout: pointer for each column
+        nDataPtrs       = nColumns;
+        nElementsPerPtr = nRows;
+    }
+    else
+    {
+        // AOS layout: one pointer for all data
+        nDataPtrs       = 1;
+        nElementsPerPtr = nRows * nColumns;
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: SDL check should be added for possible integer overflow.

File: service/kernel/data_management/finiteness_checker.cpp [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    if (inParallel)
+        daal::threader_for(nBlocks, nBlocks, [&](size_t i) { func(i); });
+    else
+        for (size_t i = 0; i < nBlocks; ++i) func(i);
+}
+
+template <typename DataType>
+DataType sumWithAVX512(size_t n, DataType * dataPtr)
+{
+    size_t nPerInstr = 64 / sizeof(DataType);
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: const

File: service/kernel/data_management/finiteness_checker.cpp [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    }
+    for (size_t i = (n / nPerInstr) * nPerInstr; i < n; ++i) sum += dataPtr[i];
+
+    return sum;
+}
+
+template <typename DataType>
+DataType computeSumAVX512Impl(size_t nDataPtrs, size_t nElementsPerPtr, DataType ** dataPtrs)
+{
+    size_t nBlocksPerPtr = nElementsPerPtr / BLOCK_SIZE;
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: const here and below

File: service/kernel/data_management/finiteness_checker.cpp [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+#endif
+
+template <typename DataType, daal::CpuType cpu>
+void allValuesAreFiniteImpl(NumericTable & table, bool allowNaN, bool * finiteness)
+{
+    const size_t nRows     = table.getNumberOfRows();
+    const size_t nColumns  = table.getNumberOfColumns();
+    const size_t nElements = nRows * nColumns;
```
Comment (ShvetsKS) [inline service/kernel/data_management/finiteness_checker.cpp]: SDL check should be added for possible integer overflow.

File: service/kernel/data_management/finiteness_checker.cpp (hunk 6 above)
Comment (ShvetsKS) [reply service/kernel/data_management/finiteness_checker.cpp]: use DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION macro for it.
see example: https://github.com/intel/daal/blob/master/algorithms/kernel/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i#L191

File: service/kernel/data_management/finiteness_checker.cpp (hunk 9 above)
Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Added

File: service/kernel/data_management/finiteness_checker.cpp (hunk 8 above)
Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Const is added where possible.

File: service/kernel/data_management/finiteness_checker.cpp (hunk 6 above)
Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Added

File: service/kernel/data_management/finiteness_checker.cpp (hunk 7 above)
Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Added

File: service/kernel/data_management/finiteness_checker.cpp (hunk 4 above)
Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Added

File: service/kernel/data_management/finiteness_checker.cpp (hunk 3 above)
Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Added

File: service/kernel/data_management/finiteness_checker.cpp (hunk 2 above)
Comment (Alexsandruss) [reply service/kernel/data_management/finiteness_checker.cpp]: Added

**Comment by Alexsandruss** [comment]
http://intel-ci.intel.com/ea514e5a-db94-f10e-a018-8cdcd4b6d2c4

**Pull request merged**